kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -484,28 +531,164 @@ private static void validateFieldCompatibility(StructField existingField, Struct
           field.getName(),
           columnId);
-      columnIdToField.put(columnId, field);
+      fieldsById.put(
+          columnId, new StructFieldWithIds(field, getColumnId(field), null, field.getName()));
+      idToFields(field, icebergCompatV2, fieldsById, null, getPhysicalName(field));
+    }
+
+    return fieldsById;
+  }
+
+  /**
+   * StructFieldWrapper wraps a StructField along with its field ID, a parent ID, and the full
+   * column path.
+   */
+  static class StructFieldWithIds implements Supplier<StructField> {
+    private final StructField field;
+    private final int id;
+    private final Integer parentId;
+    private final String columnPath;
+
+    StructFieldWithIds(StructField field, int id, Integer parentId, String columnPath) {
+      this.field = field;
+      this.id = id;
+      this.parentId = parentId;
+      this.columnPath = columnPath;
+    }
+
+    public StructField field() {
+      return field;
+    }
+
+    public int id() {
+      return id;
     }
 
-    return columnIdToField;
+    public Integer parentId() {
+      return parentId;
+    }
+
+    public String columnPath() {
+      return columnPath;
+    }
+
+    @Override
+    public StructField get() {
+      return field;
+    }
+  }
+
+  private static void idToFields(
+      StructField field,
+      boolean icebergCompatV2,
+      Map<Integer, StructFieldWithIds> idToFields,
+      Integer currentParent,
+      String fieldPath) {
+    // If the field is primitive, the mapping will already be populated
+    if (field.getDataType() instanceof BasePrimitiveType) {
+      return;
+    }
+
+    int nearestAncestorFieldId = hasColumnId(field) ? getColumnId(field) : currentParent;
+
+    if (field.getDataType() instanceof StructType) {
+      StructType structType = (StructType) field.getDataType();
+      for (StructField nestedField : structType.fields()) {
+        validateFieldHasIdAndName(nestedField);
+        int columnId = getColumnId(nestedField);
+        checkArgument(
+            !idToFields.containsKey(columnId),
+            "Field %s with id %s already exists",
+            field.getName(),
+            columnId);
+        idToFields.put(
+            columnId,
+            new StructFieldWithIds(nestedField, columnId, nearestAncestorFieldId, fieldPath));
+        idToFields(
+            nestedField,
+            icebergCompatV2,
+            idToFields,
+            nearestAncestorFieldId,
+            fieldPath + "." + getPhysicalName(nestedField));
+      }
+    } else if (field.getDataType() instanceof MapType) {
+      MapType mapType = (MapType) field.getDataType();
+      int mapKeyId;
+      int mapValueId;
+      String keyPath = fieldPath + ".key";
+      String valuePath = fieldPath + ".value";
+      if (icebergCompatV2) {
+        checkArgument(
+            hasNestedColumnIds(field),
+            "Map data types in icebergCompatV2 must have nested ID metadata defined for key and value");
```

Comment:
This is one of the keys in this PR, most of the validation that was done in https://github.com/delta-io/delta/pull/4310 is implicitly done here. The realization I had was that there's no reason to really distinguish the map.key/map.value and array.element field IDs in icebergCompatV2. We can include those in the mapping along with parent ID metadata for later checks.